### ISSUE TITLE
replace node procedure should handle unreachable seeds

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#16407]
We have recently seen the following strange case:
* Two nodes: A and B. B is down and is scheduled to be replaced.
* Bootstrapping new node C, to replace B.
* Due to bad firewall rules, packets A->C are dropped, but C->A works.
* C finishes replace procedure and creates a new cluster.
* A still thinks the cluster contains only A and B (which is down).

The expected result is: C fails bootstrap because gossiping to the seed node (A) doesn't work.

**Comment by denesb** [comment]
CC: @asias @kbr- @vladzcloudius

**Comment by vladzcloudius** [comment]
IMO the scope of this issue should be widened beyond only replace node operations: every time the node bootstraps it should make sure that important ports are accessible to/from it. AFAIU the only "important" port is a `storage_port` (7000/7001 by default).

The check can be implemented by a simple "handshake" where the node attempts to open and send an echo to the seeder and the seeder tries to do the same.
Only after that the rest of the bootstrap should continue.
If the handshake fails - the bootstrap should fail too with a corresponding and a very clear error message.

**Comment by denesb** [comment]
AFAIK, seastar RPC already has a handshake process on connection setup. Not sure how failing this didn't trigger some timeout and failing any subsequent request that is attempted to be sent. @xemul

**Comment by xemul** [comment]
If A->C doesn't work then RPC handshake (btw, it's called negotiation) shouldn't even start, because it would fail to to TCP-level connect first

**Comment by asias** [comment]
> We have recently seen the following strange case:
> 
>     * Two nodes: A and B. B is down and is scheduled to be replaced.
> 
>     * Bootstrapping new node C, to replace B.
> 
>     * Due to bad firewall rules, packets A->C are dropped, but C->A works.
> 
>     * C finishes replace procedure and creates a new cluster.
> 
>     * A still thinks the cluster contains only A and B (which is down).
> 
> 
> The expected result is: C fails bootstrap because gossiping to the seed node (A) doesn't work.

The shadow round which the replacing node gets the tokens and status uses a rpc call not gossip. That's why in this case, C knows that tokens to replace. But A could not gossip existing cluster info to C due to the network issue. A could not see other nodes in the cluster in gossip.  C thinks there are no other nodes in the cluster and continues the replace op.

**Comment by vladzcloudius** [comment]
> If A->C doesn't work then RPC handshake (btw, it's called negotiation) shouldn't even start, because it would fail to to TCP-level connect first

That's exactly what happens, @xemul, A doesn't start talking to C (see @asias comment above). However this is not the issue.
The issue is that C doesn't have a clue about that - hence the rest of the mess! 
And I suggest to implement a handshake that C will initiate which will make sure C "knows" if A can communicate with it (in addition to making sure A can communicate with C).

**Comment by xemul** [comment]
@vladzcloudius , I don't get your point. If A->C packet flow is filtered, then _neither_ A can connect to C _nor_ C can connect to A

**Comment by denesb** [comment]
I'm also confused. Can the firewall be configured such that it allows A->C traffic when the connection is initiated by C, but filter traffic for any A->C connections initiated by A? Actually, this sounds like a firewall config where outbound traffic is allowed, but inbound is filtered (except for those on already existing connections).

**Comment by mykaul** [comment]
> I'm also confused. Can the firewall be configured such that it allows A->C traffic when the connection is initiated by C, but filter traffic for any A->C connections initiated by A? Actually, this sounds like a firewall config where outbound traffic is allowed, but inbound is filtered (except for those on already existing connections).

Indeed. Any reasonable stateful firewall can allow flows from A->B, but not B->A. 

I do believe it's a rare case we shouldn't handle, btw.

**Comment by asias** [comment]
> I'm also confused. Can the firewall be configured such that it allows A->C traffic when the connection is initiated by C, but filter traffic for any A->C connections initiated by A? Actually, this sounds like a firewall config where outbound traffic is allowed, but inbound is filtered (except for those on already existing connections).


Of course. C can drop any TCP syn packages to it. This is actually what is the default. C allows only outbound traffic. That's why C an talk to A and A can return data back to C.  C does not open inbound traffic rules, So C refuses any tcp connection from A to C. 

So, I think the user forgets to open inbound security rules for C.

**Comment by denesb** [comment]
> I do believe it's a rare case we shouldn't handle, btw.

We need to detect this case and refuse finalizing boostrap. We don't need to do anything fancy: print an error and exit. I think a few strategically placed timeouts should achieve this. But we need to have tests that ensure this works.

The current result is very surprising to the operator and if they don't notice they can end up loosing data.

**Comment by xemul** [comment]
>  C does not open inbound traffic rules, So C refuses any tcp connection from A to C.

TCP establishing works like

C-[SYN]->A
A-[SYN+ACK]->C
C-[ACK]->A

If C does not open inbound traffic, why does A-[SYN-ACK]->C response passes?

**Comment by denesb** [comment]
@xemul see above. Stateful firewall allow A->C if it is in the context of a connection initiated by C. But packets are dropped if A tries to initiate connection to C.

**Comment by xemul** [comment]
Well, linux firewall is Turing machine so in theory anything is possible, but ... just to clarify

> Stateful firewall allow A->C if it is in the context of a connection initiated by C.

How about C->A in the context of a connection initiated by C?

> But packets are dropped if A tries to initiate connection to C.

Do I get it right that

* connecting A->C is allowed
  - this _implies_ that once established A->C or C->A is allowed
* connecting C->A is blocked

Is that our case?

**Comment by denesb** [comment]
AFAIK yes.

**Comment by vladzcloudius** [comment]
> OK, thanks. Then RPC handshake is not the right place to resolve this, methinks.

Where would be a good place resolve this in your opinion, @xemul ?

IMO having a timed RPC call like the one below would be totally appropriate:

```
C [rpc]-> A (executes: 
 1) create a TCP RPC connection to C
 2) A [rpc] -> C (execute timeofday))
```

If the above fails bootstrap fails with a corresponding error message.

@xemul, if you have a better idea, please, share.

**Comment by vladzcloudius** [comment]
> Well, linux firewall is Turing machine so in theory anything is possible

Assumption that this issue is related only to a Linux firewall is totally baseless.

**Comment by mykaul** [comment]
I don't think we need to resolve this. It's a rare enough misconfiguration.

**Comment by vladzcloudius** [comment]
> I don't think we need to resolve this. It's a rare enough misconfiguration.

I think we must (!!) resolve this because not resolving this silently leaves the cluster is split-brain state (==data loss) and the fix it rather trivial.

cc @tzach @ruthea @avikivity @dorlaor

**Comment by tzach** [comment]
I agree with @vladzcloudius ScyllaDB needs to (at least) identify and report split-brain state (when possible).
Can we move this loginc to and external script/tool? It will only work if it has the topology info to check connectivity.

**Comment by kbr-scylla** [comment]
You guys do realize that we're switching to raft-based topology? Such scenarios won't be possible.

> CC: @asias @kbr- @vladzcloudius

@denesb please don't tag my old account, or there's a chance I won't ever see the thread.

**Comment by denesb** [comment]
> You guys do realize that we're switching to raft-based topology? Such scenarios won't be possible.

Old releases still using gossip, will be around for years to come, especially LTS enterprise releases. So we should still consider backportable fixes/improvements that are not too difficult to implement.

> > CC: @asias @kbr- @vladzcloudius
> 
> @denesb please don't tag my old account, or there's a chance I won't ever see the thread.

Sorry, I try to keep this in mind, but sometimes github auto-complete defeats me.

**Comment by asias** [comment]
> > You guys do realize that we're switching to raft-based topology? Such scenarios won't be possible.
> 
> Old releases still using gossip, will be around for years to come, especially LTS enterprise releases. So we should still consider backportable fixes/improvements that are not too difficult to implement.
> 
> > > CC: @asias @kbr- @vladzcloudius
> > 
> > 
> > @denesb please don't tag my old account, or there's a chance I won't ever see the thread.
> 
> Sorry, I try to keep this in mind, but sometimes github auto-complete defeats me.

Or you can rename the old name.

**Comment by denesb** [comment]
Closing this as won't fix. Consistent topology changes fixes this issue and it is already used on most clusters (it is the default for new ones).
